From 5cbfb33c6eed79adc27b3db65d08c6ba0bc479ef Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Thu, 30 Jun 2022 09:16:58 +0800
Subject: [PATCH 70/97] backend-drm: Don't try to prepare non-cursor view in
 renderer-only mode

It would fail out in the later check anyway.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/backend-drm/state-propose.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/libweston/backend-drm/state-propose.c b/libweston/backend-drm/state-propose.c
index aff68e6..3141dfd 100644
--- a/libweston/backend-drm/state-propose.c
+++ b/libweston/backend-drm/state-propose.c
@@ -389,6 +389,14 @@ dmabuf_feedback_maybe_update(struct drm_device *device, struct weston_view *ev,
 	dmabuf_feedback->action_needed = ACTION_NEEDED_NONE;
 }
 
+static struct weston_layer *
+get_view_layer(struct weston_view *view)
+{
+	if (view->parent_view)
+		return get_view_layer(view->parent_view);
+	return view->layer_link.layer;
+}
+
 static struct drm_plane_state *
 drm_output_find_plane_for_view(struct drm_output_state *state,
 			       struct weston_paint_node *pnode,
@@ -404,6 +412,7 @@ drm_output_find_plane_for_view(struct drm_output_state *state,
 	struct drm_plane *plane;
 
 	struct weston_view *ev = pnode->view;
+	struct weston_layer *layer;
 	struct weston_buffer *buffer;
 	struct drm_fb *fb = NULL;
 
@@ -419,6 +428,12 @@ drm_output_find_plane_for_view(struct drm_output_state *state,
 		return NULL;
 	}
 
+	/* only allow cursor in renderer-only mode */
+	layer = get_view_layer(ev);
+	if (mode == DRM_OUTPUT_PROPOSE_STATE_RENDERER_ONLY &&
+	    layer->position != WESTON_LAYER_POSITION_CURSOR)
+		return NULL;
+
 	buffer = ev->surface->buffer_ref.buffer;
 	if (buffer->type == WESTON_BUFFER_SOLID) {
 		pnode->try_view_on_plane_failure_reasons |=
-- 
2.20.1

